stew/leb128.nim

```diff
@@ -0,0 +1,156 @@
+## more easy-to-use API.
+##
+## Exception/Defect free as of nim 1.2.
+
+{.push raises: [].}
+
+import
+  stew/bitops2
+
+const
+  # Given the truncated logarithm of a 64-bit number, how many bytes do we need
+  # to encode it?
+  lengths = block:
+    var v: array[64, int8]
+    for i in 0..<64:
+      v[i] = int8((i + 7) div 7)
+    v
+
+type
+  Leb128* = object
+    ## Type used to mark leb128 encoding helpers
+
+# log2trunc by definition never returns values >64, thus we can remove checks
+{.push checks: off.}
+func len*(T: type Leb128, x: SomeUnsignedInt): int8 =
+  ## Returns number of bytes required to encode integer ``x`` as leb128.
+  if x == 0: 1 # Always at least one byte!
+  else: lengths[log2trunc(x)]
+{.pop.}
+
+func maxLen*(T: type Leb128, I: type): int8 =
+  ## The maximum number of bytes needed to encode any value of type I
+  Leb128.len(I.high)
+
+type
+  Leb128Buf*[T: SomeUnsignedInt] = object
+    data*: array[maxLen(Leb128, T), byte] # len(data) <= 10
+    len*: int8 # >= 1 when holding valid protobuf
+
+  Leb128Error* = enum
+    incomplete = "End marker not found in given stream"
+    overflow = "Value would overflow"
+
+template write7(next: untyped) =
+  # write 7 bits of data
+  if v > type(v)(127):
+    result.data[result.len] = cast[byte](v) or 0x80'u8
+    result.len += 1
+    v = v shr 7
+    next
+
+# LebBuf size corresponds to maximum size that the type will be encoded to, thus
+# there can be no out-of-bounds accesses here - likewise with the length
+# arithmetic
+{.push checks: off.}
+func toBytes*[I: SomeUnsignedInt](v: I, T: type Leb128): Leb128Buf[I] {.noinit.} =
+  ## Convert an unsigned integer to the smallest leb128 representation possible
+  ##
+  ## Example:
+  ## 15'u16.toBytes(Leb128)
+  var
+    v = v
+  result.len = 0
+
+  # A clever developer would write something clever for the unrolling -
+  # fortunately, we have clever compilers that remove the excess unrolls based
+  # on size!
+  write7(): # 7
+    write7(): # 14
+      write7(): # 21
+        write7(): # 28
+          write7(): # 35
+            write7(): # 42
+              write7(): # 49
+                write7(): # 56
+                  write7(): # 63
+                    discard
+
+  # high bit not set since v <= 127 at this point!
+  result.data[result.len] = cast[byte](v)
+  result.len += 1
+
+template read7(shift: untyped) =
+  # Read 7 bits of data and return iff these are the last 7 bits
+  if (shift div 7) >= xlen:
+    return (I(0), 0'i8)
+  let
+    b = x[shift div 7]
+    valb = b and 0x7f'u8 # byte without high bit
+    val = I(valb)
+    vals = val shl shift
+
+  when shift >= (sizeof(val) * 8 - 7):
+    if vals shr shift != val:
+      return (I(0), -cast[int8]((shift div 7) + 1))
+
+  res = res or vals
+  if b == valb: # High bit not set
+    return (res, cast[int8]((shift div 7) + 1))
+
+func fromBytes*(
+    I: type SomeUnsignedInt,
+    x: openArray[byte],
+    T: type Leb128): tuple[val: I, len: int8] {.noinit.} =
```

Comment:
Does this decoding procedure is able to detect overlong values?
https://github.com/status-im/nim-libp2p/blob/master/tests/testvarint.nim#L382

Reply:
And how this procedure supports 9-byte length LibP2P varints?

Reply:
`leb128` doesn't have those things "natively" but it can easily be added on (which is a bit a point of these primitives):

https://github.com/status-im/nim-libp2p/pull/481/files#diff-c078cd08c868d565974266be04690ecd290f63eafaa5f51cfd4e3b9769437868R140